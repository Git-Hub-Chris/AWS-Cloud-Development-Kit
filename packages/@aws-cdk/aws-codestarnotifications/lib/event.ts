
/**
 * The list of event types for AWS Codecommit
 */
export enum RepositoryEvent {

  /**
   * Trigger notification when repository comments on commits
   */
  COMMENTS_ON_COMMITS = 'codecommit-repository-comments-on-commits',

  /**
   * Trigger notification when repository comments on pull requests
   */
  COMMENTS_ON_PULL_REQUEST = 'codecommit-repository-comments-on-pull-requests',

  /**
   * Trigger notification when repository approvals status changed
   */
  APPROVAL_STATUS_CHANGED = 'codecommit-repository-approvals-status-changed',

  /**
   * Trigger notification when repository approvals rule override
   */
  APPROVAL_RULE_OVERRIDE = 'codecommit-repository-approvals-rule-override',

  /**
   * Trigger notification when repository pull request created
   */
  PULL_REQUEST_CREATED = 'codecommit-repository-pull-request-created',

  /**
   * Trigger notification when repository pull request source updated
   */
  PULL_REQUEST_SOURCE_UPDATED = 'codecommit-repository-pull-request-source-updated',

  /**
   * Trigger notification when repository pull request status changed
   */
  PULL_REQUEST_STATUS_CHANGED = 'codecommit-repository-pull-request-status-changed',

  /**
   * Trigger notification when repository pull request merged
   */
  PULL_REQUEST_MERGED = 'codecommit-repository-pull-request-merged',

  /**
   * Trigger notification when repository branches and tags created
   */
  BRANCHES_AND_TAGS_CREATED = 'codecommit-repository-branches-and-tags-created',

  /**
   * Trigger notification when repository branches and tags deleted
   */
  BRANCHES_AND_TAGS_DELETED = 'codecommit-repository-branches-and-tags-deleted',

  /**
   * Trigger notification when repository branches and tags updated
   */
  BRANCHES_AND_TAGS_UPDATED = 'codecommit-repository-branches-and-tags-updated',
}

/**
 * The list of event types for AWS Codebuild
 */
export enum ProjectEvent {

  /**
   * Trigger notification when project build state failed
   */
  BUILD_STATE_FAILED = 'codebuild-project-build-state-failed',

  /**
   * Trigger notification when project build state succeeded
   */
  BUILD_STATE_SUCCEEDED = 'codebuild-project-build-state-succeeded',

  /**
   * Trigger notification when project build state in progress
   */
  BUILD_STATE_IN_PROGRESS = 'codebuild-project-build-state-in-progress',

  /**
   * Trigger notification when project build state stopped
   */
  BUILD_STATE_STOPPED = 'codebuild-project-build-state-stopped',

  /**
   * Trigger notification when project build phase failure
   */
  BUILD_PHASE_FAILRE = 'codebuild-project-build-phase-failure',

  /**
   * Trigger notification when project build phase success
   */
  BUILD_PHASE_SUCCESS = 'codebuild-project-build-phase-success',
}

/**
 * The list of event types for AWS Codedeploy
 */
export enum ApplicationEvent {

  /**
   * Trigger notification when application deployment failed
   */
  DEPLOYMENT_FAILED = 'codedeploy-application-deployment-failed',

  /**
   * Trigger notification when application deployment succeeded
   */
  DEPLOYMENT_SUCCEEDED = 'codedeploy-application-deployment-succeeded',

  /**
   * Trigger notification when application deployment started
   */
  DEPLOYMENT_STARTED = 'codedeploy-application-deployment-started',
}

/**
 * The list of event types for AWS Codepipeline
 */
export enum PipelineEvent {

  /**
   * Trigger notification when pipeline action execution succeeded
   */
  ACTION_EXECUTION_SUCCEEDED = 'codepipeline-pipeline-action-execution-succeeded',

  /**
   * Trigger notification when pipeline action execution failed
   */
  ACTION_EXECUTION_FAILED = 'codepipeline-pipeline-action-execution-failed',

  /**
   * Trigger notification when pipeline action execution canceled
   */
  ACTION_EXECUTION_CANCELED = 'codepipeline-pipeline-action-execution-canceled',

  /**
   * Trigger notification when pipeline action execution started
   */
  ACTION_EXECUTION_STARTED = 'codepipeline-pipeline-action-execution-started',

  /**
   * Trigger notification when pipeline stage execution started
   */
  STAGE_EXECUTION_STARTED = 'codepipeline-pipeline-stage-execution-started',

  /**
   * Trigger notification when pipeline stage execution succeeded
   */
  STAGE_EXECUTION_SUCCEEDED = 'codepipeline-pipeline-stage-execution-succeeded',

  /**
   * Trigger notification when pipeline stage execution resumed
   */
  STAGE_EXECUTION_RESUMED = 'codepipeline-pipeline-stage-execution-resumed',

  /**
   * Trigger notification when pipeline stage execution canceled
   */
  STAGE_EXECUTION_CANCELED = 'codepipeline-pipeline-stage-execution-canceled',

  /**
   * Trigger notification when pipeline stage execution failed
   */
  STAGE_EXECUTION_FAILED = 'codepipeline-pipeline-stage-execution-failed',

  /**
   * Trigger notification when pipeline execution failed
   */
  PIPELINE_EXECUTION_FAILED = 'codepipeline-pipeline-pipeline-execution-failed',

  /**
   * Trigger notification when pipeline execution canceled
   */
  PIPELINE_EXECUTION_CANCELED = 'codepipeline-pipeline-pipeline-execution-canceled',

  /**
   * Trigger notification when pipeline execution started
   */
  PIPELINE_EXECUTION_STARTED = 'codepipeline-pipeline-pipeline-execution-started',

  /**
   * Trigger notification when pipeline execution resumed
   */
  PIPELINE_EXECUTION_RESUMED = 'codepipeline-pipeline-pipeline-execution-resumed',

  /**
   * Trigger notification when pipeline execution succeeded
   */
  PIPELINE_EXECUTION_SUCCEEDED = 'codepipeline-pipeline-pipeline-execution-succeeded',

  /**
   * Trigger notification when pipeline execution superseded
   */
  PIPELINE_EXECUTION_SUPERSEDED = 'codepipeline-pipeline-pipeline-execution-superseded',

  /**
   * Trigger notification when pipeline manual approval failed
   */
  MANUAL_APPROVAL_FAILED = 'codepipeline-pipeline-manual-approval-failed',

  /**
   * Trigger notification when pipeline manual approval needed
   */
  MANUAL_APPROVAL_NEEDED = 'codepipeline-pipeline-manual-approval-needed',

  /**
   * Trigger notification when pipeline manual approval succeeded
   */
  MANUAL_APPROVAL_SUCCEEDED = 'codepipeline-pipeline-manual-approval-succeeded',
}