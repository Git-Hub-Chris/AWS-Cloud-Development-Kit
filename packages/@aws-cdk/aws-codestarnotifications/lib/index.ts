// AWS::CodeStarNotifications CloudFormation Resources:
export * from './codestarnotifications.generated';
export * from './notification-rule';
export * from './target';
export * from './event';