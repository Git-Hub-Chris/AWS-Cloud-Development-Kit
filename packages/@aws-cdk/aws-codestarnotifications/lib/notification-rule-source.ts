
import * as constructs from 'constructs';
import { INotificationRule } from './notification-rule';

/**
 * Information about the Codebuild or CodePipeline associated with a notification source.
 */
export interface NotificationRuleSourceConfig {
  /**
   * The source type. Can be an AWS CodeCommit, CodeBuild, CodePipeline or CodeDeploy.
   */
  readonly sourceType: string;

  /**
   * The Amazon Resource Name (ARN) of the notification source.
   */
  readonly sourceArn: string;
}

/**
 * Represents a notification source
 * The source that allows CodeBuild and CodePipeline to associate with this rule.
 */
export interface INotificationRuleSource {
  /**
   * Returns a source configuration for notification rule.
   */
  bindAsNotificationRuleSource(scope: constructs.Construct, rule: INotificationRule): NotificationRuleSourceConfig;
}
