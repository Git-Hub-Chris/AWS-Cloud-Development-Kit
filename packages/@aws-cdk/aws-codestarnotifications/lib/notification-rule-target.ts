
import { INotificationRule } from './notification-rule';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from '@aws-cdk/core';

/**
 * Information about the SNS topic or AWS Chatbot client associated with a notification target.
 */
export interface NotificationRuleTargetConfig {
  /**
   * The target type. Can be an Amazon SNS topic or AWS Chatbot client.
   */
  readonly targetType: string;

  /**
   * The Amazon Resource Name (ARN) of the Amazon SNS topic or AWS Chatbot client.
   */
  readonly targetAddress: string;
}

/**
 * Represents a notification target
 * That allows AWS Chatbot and SNS topic to associate with this rule target.
 */
export interface INotificationRuleTarget {
  /**
   * Returns a target configuration for notification rule.
   */
  bindAsNotificationRuleTarget(scope: Construct, rule: INotificationRule): NotificationRuleTargetConfig;
}