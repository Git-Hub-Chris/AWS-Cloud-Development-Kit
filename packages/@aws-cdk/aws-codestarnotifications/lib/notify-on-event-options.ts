import { Status, DetailType } from './rule';
import { IRuleTarget } from './target';

/**
 * Standard set of options for `notifyOnXxx` codestar notification handler on construct
 */
export interface NotifyOnEventOptions {
  /**
   * The name for the notification rule.
   * Notification rule names must be unique in your AWS account.
   *
   * @default - generated from the `id`
   */
  readonly ruleName?: string;

  /**
   * The target to register for the notification destination.
   *
   * @default - No target is added to the rule. Use `addTarget()` to add a target.
   */
  readonly target?: IRuleTarget;

  /**
   * A list of event types associated with this notification rule.
   * For a complete list of event types and IDs, see Notification concepts in the Developer Tools Console User Guide.
   * @see https://docs.aws.amazon.com/dtconsole/latest/userguide/concepts.html#concepts-api
   *
   * You must specify this property (either via props or via `addEvents`)
   *
   * @default - No events.
   */
  readonly events?: string[];

  /**
   * The status of the notification rule.
   * If the status is set to DISABLED, notifications aren't sent for the notification rule.
   *
   * @default Status.ENABLED
   */
  readonly status?: Status;

  /**
   * The level of detail to include in the notifications for this resource.
   * BASIC will include only the contents of the event as it would appear in AWS CloudWatch.
   * FULL will include any supplemental information provided by AWS CodeStar Notifications and/or the service for the resource for which the notification is created.
   *
   * @default DetailType.FULL
   */
  readonly detailType?: DetailType;
}