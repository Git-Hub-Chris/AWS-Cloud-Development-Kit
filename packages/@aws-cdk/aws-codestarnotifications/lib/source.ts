import { INotificationRule } from './notification-rule';

/**
 * Information about the Codebuild or CodePipeline associated with a notification source.
 */
export interface NotificationRuleSourceConfig {
  /**
   * The source type. Can be an AWS CodeCommit, CodeBuild, CodePipeline or CodeDeploy.
   */
  readonly sourceType: string;

  /**
   * The Amazon Resource Name (ARN) of the notification source.
   */
  readonly sourceArn: string;
}

/**
 * Represents a notification source
 * The source that allows CodeBuild and CodePipeline to associate with this rule.
 */
export interface INotificationRuleSource {
  /**
   * Binds source to notification rule
   * @param _rule The notification rule
   */
  bind(_rule: INotificationRule): NotificationRuleSourceConfig;
}