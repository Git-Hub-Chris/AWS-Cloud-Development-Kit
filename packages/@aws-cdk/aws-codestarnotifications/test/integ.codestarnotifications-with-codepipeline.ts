import * as chatbot from '@aws-cdk/aws-chatbot';
import * as codebuild from '@aws-cdk/aws-codebuild';
import * as codecommit from '@aws-cdk/aws-codecommit';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipelineActions from '@aws-cdk/aws-codepipeline-actions';
import * as s3 from '@aws-cdk/aws-s3';
import * as sns from '@aws-cdk/aws-sns';
import * as cdk from '@aws-cdk/core';
import { NotificationRule, SlackNotificationTarget, SNSTopicNotificationTarget, Status, DetailType, PipelineEvent } from '../lib';

class CodestarnotificationsWithCodepipelineInteg extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const repository = new codecommit.Repository(this, 'MyRepo', {
      repositoryName: 'my-repo',
    });

    const sourceOutput = new codepipeline.Artifact();

    const artifactBucket = new s3.Bucket(this, 'MyBucket', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const pipeline = new codepipeline.Pipeline(this, 'MyPipeline', {
      artifactBucket: artifactBucket,
      stages: [
        {
          stageName: 'SourceStage',
          actions: [
            new codepipelineActions.CodeCommitSourceAction({
              actionName: 'FetchSourceChanged',
              repository: repository,
              output: sourceOutput,
            }),
          ],
        },
        {
          stageName: 'TestStage',
          actions: [
            new codepipelineActions.CodeBuildAction({
              actionName: 'Execute',
              input: sourceOutput,
              project: new codebuild.PipelineProject(this, 'MyTestCodebuildProject', {
                buildSpec: codebuild.BuildSpec.fromObject({
                  version: '0.2',
                  phases: {
                    install: {
                      'runtime-versions': {
                        docker: 18,
                      },
                    },
                    build: {
                      commands: [
                        'echo "Nothing to do!"',
                      ],
                    },
                  },
                }),
              }),
            }),
          ],
        },
      ],
    });

    const slackChannel = new chatbot.SlackChannelConfiguration(this, 'MySlackChannel', {
      slackChannelConfigurationName: 'test-slack-with-codebuild',
      slackWorkspaceId: 'T49239U4W',
      slackChannelId: 'C0187JABUE9',
    });

    const topic1 = new sns.Topic(this, 'MyTopic1', {});
    const topic2 = new sns.Topic(this, 'MyTopic2', {});

    const notificationRule = new NotificationRule(this, 'MyNotificationRule', {
      notificationRuleName: 'MyNotificationRule',
      status: Status.ENABLED,
      detailType: DetailType.FULL,
      eventTypeIds: [
        PipelineEvent.PIPELINE_EXECUTION_SUCCEEDED,
        PipelineEvent.PIPELINE_EXECUTION_FAILED,
      ],
      targets: [
        new SlackNotificationTarget(slackChannel),
        new SNSTopicNotificationTarget(topic1),
      ],
      resourceArn: pipeline.pipelineArn,
    });

    notificationRule.addTarget(new SNSTopicNotificationTarget(topic2));
  }
}

const app = new cdk.App();

new CodestarnotificationsWithCodepipelineInteg(app, 'CodestarnotificationsWithCodepipelineInteg');

app.synth();