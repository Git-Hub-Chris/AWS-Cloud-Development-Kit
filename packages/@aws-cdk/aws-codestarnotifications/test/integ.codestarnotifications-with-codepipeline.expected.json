{
  "Resources": {
    "MyRepoF4F48043": {
      "Type": "AWS::CodeCommit::Repository",
      "Properties": {
        "RepositoryName": "my-repo"
      }
    },
    "MyRepoCodestarnotificationsWithCodepipelineIntegMyPipelineAEC7785AEventRuleEAF41145": {
      "Type": "AWS::Events::Rule",
      "Properties": {
        "EventPattern": {
          "source": [
            "aws.codecommit"
          ],
          "resources": [
            {
              "Fn::GetAtt": [
                "MyRepoF4F48043",
                "Arn"
              ]
            }
          ],
          "detail-type": [
            "CodeCommit Repository State Change"
          ],
          "detail": {
            "event": [
              "referenceCreated",
              "referenceUpdated"
            ],
            "referenceName": [
              "master"
            ]
          }
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition"
                  },
                  ":codepipeline:",
                  {
                    "Ref": "AWS::Region"
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId"
                  },
                  ":",
                  {
                    "Ref": "MyPipelineAED38ECF"
                  }
                ]
              ]
            },
            "Id": "Target0",
            "RoleArn": {
              "Fn::GetAtt": [
                "MyPipelineEventsRoleFAB99F32",
                "Arn"
              ]
            }
          }
        ]
      }
    },
    "MyBucketF68F3FF0": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "MyTestCodebuildProjectRole0B4E82E4": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyTestCodebuildProjectRoleDefaultPolicy723A381D": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "MyTestCodebuildProject7959878A"
                      }
                    ]
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "MyTestCodebuildProject7959878A"
                      },
                      ":*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":codebuild:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":report-group/",
                    {
                      "Ref": "MyTestCodebuildProject7959878A"
                    },
                    "-*"
                  ]
                ]
              }
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyTestCodebuildProjectRoleDefaultPolicy723A381D",
        "Roles": [
          {
            "Ref": "MyTestCodebuildProjectRole0B4E82E4"
          }
        ]
      }
    },
    "MyTestCodebuildProject7959878A": {
      "Type": "AWS::CodeBuild::Project",
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE"
        },
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:1.0",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER"
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "MyTestCodebuildProjectRole0B4E82E4",
            "Arn"
          ]
        },
        "Source": {
          "BuildSpec": "{\n  \"version\": \"0.2\",\n  \"phases\": {\n    \"install\": {\n      \"runtime-versions\": {\n        \"docker\": 18\n      }\n    },\n    \"build\": {\n      \"commands\": [\n        \"echo \\\"Nothing to do!\\\"\"\n      ]\n    }\n  }\n}",
          "Type": "CODEPIPELINE"
        }
      }
    },
    "MyPipelineRoleC0D47CA4": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyPipelineRoleDefaultPolicy34F09EFA": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyPipelineSourceStageFetchSourceChangedCodePipelineActionRoleE83DDD81",
                  "Arn"
                ]
              }
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyPipelineTestStageExecuteCodePipelineActionRole2BB9A57E",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyPipelineRoleDefaultPolicy34F09EFA",
        "Roles": [
          {
            "Ref": "MyPipelineRoleC0D47CA4"
          }
        ]
      }
    },
    "MyPipelineAED38ECF": {
      "Type": "AWS::CodePipeline::Pipeline",
      "Properties": {
        "RoleArn": {
          "Fn::GetAtt": [
            "MyPipelineRoleC0D47CA4",
            "Arn"
          ]
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeCommit",
                  "Version": "1"
                },
                "Configuration": {
                  "RepositoryName": {
                    "Fn::GetAtt": [
                      "MyRepoF4F48043",
                      "Name"
                    ]
                  },
                  "BranchName": "master",
                  "PollForSourceChanges": false
                },
                "Name": "FetchSourceChanged",
                "OutputArtifacts": [
                  {
                    "Name": "Artifact_SourceStage_FetchSourceChanged"
                  }
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "MyPipelineSourceStageFetchSourceChangedCodePipelineActionRoleE83DDD81",
                    "Arn"
                  ]
                },
                "RunOrder": 1
              }
            ],
            "Name": "SourceStage"
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1"
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "MyTestCodebuildProject7959878A"
                  }
                },
                "InputArtifacts": [
                  {
                    "Name": "Artifact_SourceStage_FetchSourceChanged"
                  }
                ],
                "Name": "Execute",
                "RoleArn": {
                  "Fn::GetAtt": [
                    "MyPipelineTestStageExecuteCodePipelineActionRole2BB9A57E",
                    "Arn"
                  ]
                },
                "RunOrder": 1
              }
            ],
            "Name": "TestStage"
          }
        ],
        "ArtifactStore": {
          "Location": {
            "Ref": "MyBucketF68F3FF0"
          },
          "Type": "S3"
        }
      },
      "DependsOn": [
        "MyPipelineRoleDefaultPolicy34F09EFA",
        "MyPipelineRoleC0D47CA4"
      ]
    },
    "MyPipelineSourceStageFetchSourceChangedCodePipelineActionRoleE83DDD81": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyPipelineSourceStageFetchSourceChangedCodePipelineActionRoleDefaultPolicyE69235AD": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "codecommit:GetBranch",
                "codecommit:GetCommit",
                "codecommit:UploadArchive",
                "codecommit:GetUploadArchiveStatus",
                "codecommit:CancelUploadArchive"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyRepoF4F48043",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyPipelineSourceStageFetchSourceChangedCodePipelineActionRoleDefaultPolicyE69235AD",
        "Roles": [
          {
            "Ref": "MyPipelineSourceStageFetchSourceChangedCodePipelineActionRoleE83DDD81"
          }
        ]
      }
    },
    "MyPipelineEventsRoleFAB99F32": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyPipelineEventsRoleDefaultPolicyF045F033": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codepipeline:StartPipelineExecution",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":codepipeline:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":",
                    {
                      "Ref": "MyPipelineAED38ECF"
                    }
                  ]
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyPipelineEventsRoleDefaultPolicyF045F033",
        "Roles": [
          {
            "Ref": "MyPipelineEventsRoleFAB99F32"
          }
        ]
      }
    },
    "MyPipelineTestStageExecuteCodePipelineActionRole2BB9A57E": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyPipelineTestStageExecuteCodePipelineActionRoleDefaultPolicyF9F3B765": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyTestCodebuildProject7959878A",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyPipelineTestStageExecuteCodePipelineActionRoleDefaultPolicyF9F3B765",
        "Roles": [
          {
            "Ref": "MyPipelineTestStageExecuteCodePipelineActionRole2BB9A57E"
          }
        ]
      }
    },
    "MySlackChannelConfigurationRole1D3F23AE": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "chatbot.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MySlackChannelA8E0B56C": {
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
      "Properties": {
        "ConfigurationName": "test-slack-with-codebuild",
        "IamRoleArn": {
          "Fn::GetAtt": [
            "MySlackChannelConfigurationRole1D3F23AE",
            "Arn"
          ]
        },
        "SlackChannelId": "C0187JABUE9",
        "SlackWorkspaceId": "T49239U4W",
        "LoggingLevel": "NONE"
      }
    },
    "MyTopic13BD94FE8": {
      "Type": "AWS::SNS::Topic"
    },
    "MyTopic1PolicyE867962F": {
      "Type": "AWS::SNS::TopicPolicy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": {
                "Service": "codestar-notifications.amazonaws.com"
              },
              "Resource": {
                "Ref": "MyTopic13BD94FE8"
              },
              "Sid": "0"
            }
          ],
          "Version": "2012-10-17"
        },
        "Topics": [
          {
            "Ref": "MyTopic13BD94FE8"
          }
        ]
      }
    },
    "MyTopic288CE2107": {
      "Type": "AWS::SNS::Topic"
    },
    "MyTopic2PolicyEE5C0174": {
      "Type": "AWS::SNS::TopicPolicy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": {
                "Service": "codestar-notifications.amazonaws.com"
              },
              "Resource": {
                "Ref": "MyTopic288CE2107"
              },
              "Sid": "0"
            }
          ],
          "Version": "2012-10-17"
        },
        "Topics": [
          {
            "Ref": "MyTopic288CE2107"
          }
        ]
      }
    },
    "MyNotificationRuleE8F95AAB": {
      "Type": "AWS::CodeStarNotifications::NotificationRule",
      "Properties": {
        "DetailType": "FULL",
        "EventTypeIds": [
          "codepipeline-pipeline-pipeline-execution-succeeded",
          "codepipeline-pipeline-pipeline-execution-failed"
        ],
        "Name": "MyNotificationRule",
        "Resource": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":codepipeline:",
              {
                "Ref": "AWS::Region"
              },
              ":",
              {
                "Ref": "AWS::AccountId"
              },
              ":",
              {
                "Ref": "MyPipelineAED38ECF"
              }
            ]
          ]
        },
        "Targets": [
          {
            "TargetAddress": {
              "Ref": "MySlackChannelA8E0B56C"
            },
            "TargetType": "AWSChatbotSlack"
          },
          {
            "TargetAddress": {
              "Ref": "MyTopic13BD94FE8"
            },
            "TargetType": "SNS"
          },
          {
            "TargetAddress": {
              "Ref": "MyTopic288CE2107"
            },
            "TargetType": "SNS"
          }
        ],
        "Status": "ENABLED"
      }
    }
  },
  "Outputs": {
    "Test": {
      "Value": "luckily"
    }
  }
}